File: LeetCode/01372. Longest ZigZag Path in a Binary Tree/01372.c
// https://leetcode.com/problems/longest-zigzag-path-in-a-binary-tree

#define L 1
#define R 2

int longest;

void helper(struct TreeNode *root, int from, int count) {
    if (root == NULL) {
        if (longest < count) {
            longest = count;
        }
        return;
    }
    
    if (from == L) {
        helper(root->left, L, 0);
        helper(root->right, R, count+1);
    } else {
        helper(root->left, L, count+1);
        helper(root->right, R, 0);
    }
}

int longestZigZag(struct TreeNode *root) {
    int ans;

    helper(root->left, L, 0);
    helper(root->right, R, 0);
    
    ans = longest;
    longest = 0;
    return ans;
}

File: LeetCode/01372. Longest ZigZag Path in a Binary Tree/01372.go
// https://leetcode.com/problems/longest-zigzag-path-in-a-binary-tree

func longestZigZag(root *TreeNode) int {
    const (
        LEFT = iota
        RIGHT
    )

    ans := 0

    var dfs func(node *TreeNode, goLeft bool, level int)
    dfs = func(node *TreeNode, goLeft bool, level int) {
        if node == nil {
            if ans < level {
                ans = level
            }
            return
        }

        if goLeft {
            dfs(node.Left, false, level+1)
            dfs(node.Right, true, 0)
        } else {
            dfs(node.Left, false, 0)
            dfs(node.Right, true, level+1)
        }
    }

    dfs(root.Left, false, 0)
    dfs(root.Right, true, 0)

    return ans
}
